tests/cxx/implicit_ctor-d1.h

```diff
@@ -21,15 +22,44 @@ int ImplicitCtorFn(IndirectWithImplicitCtor);
 // IWYU: IndirectWithImplicitCtor is...*implicit_ctor-i2.h.*for autocast
 int ImplicitCtorRefFn(const IndirectWithImplicitCtor&);
 
+// Autocast assurance for header-defined functions still makes sense as opposed
+// to function definitions in source files.
+// IWYU: IndirectWithImplicitCtor needs a declaration
+// IWYU: IndirectWithImplicitCtor is...*implicit_ctor-i2.h.*for autocast
+inline int InlineImplicitCtorRefFn(const IndirectWithImplicitCtor&) {
+  return 1;
+}
+
+// IndirectClass hasn't any implicit, one-argument constructor, so fwd declaring
+// is ok.
+int NoAutocastFn(
+    // IWYU: IndirectClass needs a declaration
+    const IndirectClass&,
+    // A subtle c++ point: forward-declaring is ok for nonconst, because
+    // you can't do implicit conversion to a non-const reference
+    // (implicit conversion involves creating a temporary, which
+    // doesn't bind to non-const references).
+    // IWYU: IndirectWithImplicitCtor needs a declaration
+    IndirectWithImplicitCtor& nonconst,
+    // Forward-declaring is ok because ptrref is a const reference
+    // to a *pointer*.
+    // IWYU: IndirectWithImplicitCtor needs a declaration
+    IndirectWithImplicitCtor* const& ptrref,
+    // IWYU: IndirectClass needs a declaration
+    IndirectClass);
```

Comment:
Maybe, add "autocastable" parameter in order to test mixing it with non-"autocastable", like it was in `badinc`? Or, `iwyu_stricter_than_cpp` covers already that case?..

Reply:
I'm not sure if that's explicitly covered, but I don't think it adds a lot of value. Possibly as a second function that just tests that interaction:
```
void AutocastAndNot(
    // IWYU: IndirectWithImplicitCtor needs a declaration
    // IWYU: IndirectWithImplicitCtor is...*implicit_ctor-i2.h.*for autocast
    const IndirectWIthImplicitCtor&,
    // IndirectClass does not have any implicit constructors.
    // IWYU: IndirectClass needs a declaration
    const IndirectClass&
);
```
But it feels like that's a testcase we can add if someone reports a problem :-)